README.md
# Liaison Installer for CodeIgniter4

![QA Control](https://github.com/paulbalandan/codeigniter4-installer/workflows/QA%20Control/badge.svg?branch=develop)
[![GitHub license](https://img.shields.io/github/license/paulbalandan/codeigniter4-installer)](LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg)](https://github.com/paulbalandan/codeigniter4-installer/pulls)

## Installation

Using Composer, globally install this Installer using the following command:
```bash
    composer global require paulbalandan/codeigniter4-installer
```

## Usage

Now that you have globally installed the Installer, you can just use `codeigniter4 new [name]` in your terminal, where `name` is the name of the directory, to call the scaffolding.

`name` here is optional. If you did not provide one, this will default to your current working directory.
If `name` is provided, it will be resolved relative to the current working directory.

Run `codeigniter4 new --help` for the full options.

## Options

- `--dev` - Installs the latest CI4 developer version instead of the latest stable release
- `--with-git` - Initializes an empty Git repository in the directory
- `--with-gitflow` - Uses GitFlow to initialize the Git repository. This has `--with-git` option implicitly included.
- `-f|--force` - Force install on existing directory.

**Note:** These options are not enabled by default. You should provide your set of options.


CHANGELOG.md
# Changelog

## [v1.0.0](https://github.com/paulbalandan/codeigniter4-installer/tree/v1.0.0) - 2020-07-15

* Initial release


tests/Command/NewCommandTest.php
<?php

/**
 * This file is part of Liaison Installer for CodeIgniter4.
 *
 * (c) John Paul E. Balandan, CPA <paulbalandan@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Liaison\Installers\CodeIgniter4\Tests\Command;

use Liaison\Installers\CodeIgniter4\Command\NewCommand;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Tester\CommandTester;
use Symfony\Component\Filesystem\Filesystem;

/**
 * NewCommandTest
 *
 * @author John Paul E. Balandan <paulbalandan@gmail.com>
 */
class NewCommandTest extends TestCase
{
    /** @var Filesystem */
    protected $fs;

    /** @var string */
    protected $name;

    /** @var string */
    protected $directory = '';

    protected function setUp(): void
    {
        $this->fs        = new Filesystem();
        $this->name      = 'output/app';
        $this->directory = __DIR__ . '/../../output/app';

        if ($this->fs->exists($this->directory)) {
            $this->fs->chmod($this->directory, 0777, 0000, true);
            $this->fs->remove($this->directory);
        }
    }

    protected function tearDown(): void
    {
        if ($this->fs->exists($this->directory)) {
            $this->fs->chmod($this->directory, 0777, 0000, true);
            $this->fs->remove($this->directory);
        }
    }

    public function testInstallerCanScaffoldANewCodeigniterApp()
    {
        $app = new Application('Liaison Installer for CodeIgniter4');
        $app->add(new NewCommand());

        $tester = new CommandTester($app->find('new'));

        $exitCode = $tester->execute(['name' => $this->name, '-v' => null]);

        $this->assertEquals(0, $exitCode);
        $this->assertDirectoryExists($this->directory . '/vendor');
        $this->assertFileExists($this->directory . '/.env');
    }

    public function testInstallerCanScaffoldADeveloperCodeigniterApp()
    {
        $app = new Application('Liaison Installer for CodeIgniter4');
        $app->add(new NewCommand());

        $tester = new CommandTester($app->find('new'));

        $exitCode = $tester->execute(['name' => $this->name, '--dev' => null, '-v' => null]);

        $this->assertEquals(0, $exitCode);
        $this->assertDirectoryExists($this->directory . '/vendor');
        $this->assertDirectoryNotExists($this->directory . '/admin');
        $this->assertFileNotExists($this->directory . '/CONTRIBUTING.md');
    }

    public function testInstallerCanInitializeGit()
    {
        $app = new Application('Liaison Installer for CodeIgniter4');
        $app->add(new NewCommand());

        $tester = new CommandTester($app->find('new'));

        $exitCode = $tester->execute(['name' => $this->name, '--with-git' => null, '-v' => null]);

        $this->assertEquals(0, $exitCode);
        $this->assertDirectoryExists($this->directory . '/.git');
    }
}


.rules.php
<?php

/**
 * This file is part of Liaison Installer for CodeIgniter4.
 *
 * (c) John Paul E. Balandan, CPA <paulbalandan@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

return [
    '@PSR2'                                            => true, // switch to PSR12 when available
    'array_syntax'                                     => ['syntax' => 'short'],
    'backtick_to_shell_exec'                           => true,
    'binary_operator_spaces'                           => ['default' => 'align_single_space'],
    'blank_line_after_opening_tag'                     => true,
    'cast_spaces'                                      => true,
    'combine_consecutive_unsets'                       => true,
    'compact_nullable_typehint'                        => true,
    'concat_space'                                     => ['spacing' => 'one'],
    'explicit_indirect_variable'                       => true,
    'function_to_constant'                             => ['functions' => ['get_called_class', 'get_class', 'get_class_this', 'php_sapi_name', 'phpversion', 'pi']],
    'function_typehint_space'                          => true,
    'include'                                          => true,
    'increment_style'                                  => ['style' => 'post'],
    'linebreak_after_opening_tag'                      => true,
    'list_syntax'                                      => ['syntax' => 'short'],
    'logical_operators'                                => true,
    'lowercase_cast'                                   => true,
    'lowercase_static_reference'                       => true,
    'mb_str_functions'                                 => true,
    'method_chaining_indentation'                      => true,
    'modernize_types_casting'                          => true,
    'multiline_comment_opening_closing'                => true,
    'multiline_whitespace_before_semicolons'           => ['strategy' => 'new_line_for_chained_calls'],
    'native_function_invocation'                       => ['include' => ['@compiler_optimized'], 'scope' => 'namespaced', 'strict' => true],
    'new_with_braces'                                  => true,
    'no_blank_lines_after_class_opening'               => true,
    'no_blank_lines_after_phpdoc'                      => true,
    'no_trailing_comma_in_singleline_array'            => true,
    'no_whitespace_in_blank_line'                      => true,
    'no_spaces_around_offset'                          => true,
    'no_unused_imports'                                => true,
    'no_useless_else'                                  => true,
    'no_useless_return'                                => true,
    'no_whitespace_before_comma_in_array'              => true,
    'nullable_type_declaration_for_default_null_value' => true,
    'phpdoc_add_missing_param_annotation'              => ['only_untyped' => false],
    'phpdoc_order'                                     => true,
    'phpdoc_scalar'                                    => true,
    'phpdoc_var_without_name'                          => true,
    'pow_to_exponentiation'                            => true,
    'psr4'                                             => true,
    'random_api_migration'                             => true,
    'return_type_declaration'                          => true,
    'short_scalar_cast'                                => true,
    'simplified_null_return'                           => false,
    'single_blank_line_before_namespace'               => true,
    'single_quote'                                     => true,
    'yoda_style'                                       => true,
];


src/Command/NewCommand.php
<?php

/**
 * This file is part of Liaison Installer for CodeIgniter4.
 *
 * (c) John Paul E. Balandan, CPA <paulbalandan@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Liaison\Installers\CodeIgniter4\Command;

use GuzzleHttp\Client;
use RuntimeException;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Filesystem\Exception\IOExceptionInterface;
use Symfony\Component\Filesystem\Filesystem;
use Symfony\Component\Process\ExecutableFinder;
use Symfony\Component\Process\PhpExecutableFinder;
use Symfony\Component\Process\Process;
use ZipArchive;

/**
 * NewCommand
 *
 * Scaffolds an entire CodeIgniter4 application.
 *
 * @author John Paul E. Balandan, CPA <paulbalandan@gmail.com>
 */
class NewCommand extends Command
{
    /**
     * Git config variables
     *
     * @var array
     */
    protected $gitConfig = [];

    /**
     * Filesystem object
     *
     * @var \Symfony\Component\Filesystem\Filesystem
     */
    protected $fs;

    /**
     * InputInterface object
     *
     * @var \Symfony\Component\Console\Input\InputInterface
     */
    protected $input;

    /**
     * OutputInterface object
     *
     * @var \Symfony\Component\Console\Output\OutputInterface
     */
    protected $output;

    /**
     * Configures the current command.
     *
     * @return void
     */
    protected function configure()
    {
        $this
            ->setName('new')
            ->setDescription('Create a new CodeIgniter4 application.')
            ->addArgument('name', InputArgument::OPTIONAL, 'Name of the local directory where the application will be made.')
            ->addOption('dev', null, InputOption::VALUE_NONE, 'Installs the latest CI4 developer version')
            ->addOption('with-git', null, InputOption::VALUE_NONE, 'Initializes an empty Git repository in the directory.')
            ->addOption('with-gitflow', null, InputOption::VALUE_NONE, 'Uses GitFlow to initialize the Git repository. This has "--with-git" option implicitly included.')
            ->addOption('force', 'f', InputOption::VALUE_NONE, 'Force install on existing directory.')
        ;
    }

    /**
     * Execute the command.
     *
     * @param \Symfony\Component\Console\Input\InputInterface   $input
     * @param \Symfony\Component\Console\Output\OutputInterface $output
     *
     * @return int
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        // save these for use by other methods later
        $this->fs     = new Filesystem();
        $this->input  = $input;
        $this->output = $output;

        if (!\extension_loaded('zip')) {
            $this->output->writeln('<error>Liaison Installer for CodeIgniter4 needs the ZIP extension installed.</error>');
            return 1;
        }

        $name      = $this->input->getArgument('name');
        $directory = ($name && '.' !== $name) ? getcwd() . DIRECTORY_SEPARATOR . $name : getcwd();

        if (!$this->input->getOption('force') && $this->verifyApplicationDirectory($directory)) {
            $this->output->writeln('<error>Application already exists!</error>');
            return 1;
        }

        $this->output->writeln('<info>Creating your own CodeIgniter4 application...</info>');
        $zipFile = $this->getFilename();

        return $this
            ->download($zipFile, $this->getVersion())
            ->extract($zipFile, $directory)
            ->removeExtraneousFiles($directory)
            ->createFiles($directory)
            ->prepareWritableDirectory($directory)
            ->configureSystemPath($directory)
            ->cleanUp($zipFile)
            ->prepareComposerJson($directory)
            ->initializeGit($directory)
            ->initializeGitFlow($directory)
            ->installApplication($directory)
        ;
    }

    /**
     * Verify that the application directory does not exist.
     *
     * @param string $directory
     *
     * @return bool
     */
    protected function verifyApplicationDirectory(string $directory)
    {
        return $this->fs->exists($directory) && $directory !== getcwd();
    }

    /**
     * Gets the temporary filename for the zip file.
     *
     * @return string
     */
    protected function getFilename(): string
    {
        return $this->fs->tempnam(getcwd(), 'codeigniter4_', '.zip');
    }

    /**
     * Gets the CI4 version to download.
     *
     * @return string
     */
    protected function getVersion(): string
    {
        if ($this->input->getOption('dev')) {
            return 'develop';
        }

        return 'appstarter';
    }

    /**
     * Gets the zipball URL for the latest release of appstarter.
     *
     * @throws RuntimeException
     * @return string
     */
    protected function getAppstarterURL(): string
    {
        $response = (new Client())->get('https://api.github.com/repos/codeigniter4/appstarter/releases/latest');

        $json = json_decode($response->getBody(), true);

        if (!isset($json['zipball_url'])) {
            throw new RuntimeException('Cannot determine the ZIP URL from GitHub API.');
        }

        return $json['zipball_url'];
    }

    /**
     * Downloads the zip file.
     *
     * @param string $zipFile
     * @param string $version
     *
     * @throws \GuzzleHTTP\Exception\GuzzleException
     * @throws \Symfony\Component\Filesystem\Exception\IOExceptionInterface
     * @return $this
     */
    protected function download(string $zipFile, string $version = 'framework')
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Downloading the zip file...</info>');
        }

        switch ($version) {
            case 'develop':
                $zip = 'https://github.com/codeigniter4/CodeIgniter4/zipball/develop';
                break;
            case 'appstarter':
            default:
                $zip = $this->getAppstarterURL();
                break;
        }

        $response = (new Client())->get($zip);

        $this->fs->dumpFile($zipFile, $response->getBody());
        return $this;
    }

    /**
     * Extracts the zip file into the directory
     *
     * @param string $zipFile
     * @param string $directory
     *
     * @throws RuntimeException
     * @return $this
     */
    protected function extract(string $zipFile, string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Extracting zip file...</info>');
        }

        $archive  = new ZipArchive();
        $response = $archive->open($zipFile, ZipArchive::CHECKCONS);

        if (!$response) {
            throw new RuntimeException('The zip file errored during extraction.');
        }

        // the zip has a parent folder so we need to know its name
        // to properly extract the subdirectories
        $parentFolder = $archive->getNameIndex(0);

        $archive->extractTo($directory);
        $archive->close();

        // move the subdirs one folder up
        $this->fs->mirror($directory . DIRECTORY_SEPARATOR . $parentFolder, $directory);
        $this->fs->remove($directory . DIRECTORY_SEPARATOR . $parentFolder);

        return $this;
    }

    /**
     * Removes extraneous files and directories.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function removeExtraneousFiles(string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Removing extraneous files and directories...</info>');
        }

        $dir = $directory . DIRECTORY_SEPARATOR;

        // common deletables in appstarter and develop
        $extras = [
            $dir . '.github',
            $dir . 'tests',
            $dir . 'README.md',
            $dir . 'phpunit.xml.dist',
        ];

        if ($this->input->getOption('dev')) {
            $extras = array_merge($extras, [
                $dir . 'admin',
                $dir . 'contributing',
                $dir . 'system',
                $dir . 'user_guide_src',
                $dir . '.editorconfig',
                $dir . '.nojekyll',
                $dir . '.travis.yml',
                $dir . 'CHANGELOG.md',
                $dir . 'CODE_OF_CONDUCT.md',
                $dir . 'CONTRIBUTING.md',
                $dir . 'DCO.txt',
                $dir . 'phpdoc.dist.xml',
                $dir . 'PULL_REQUEST_TEMPLATE.md',
                $dir . 'stale.yml',
                $dir . 'Vagrantfile.dist',
            ]);
        }

        try {
            $this->fs->remove($extras);
        } catch (IOExceptionInterface $e) {
            $this->output->writeln('<error>' . $e->getMessage() . '</error>');
        }

        return $this;
    }

    /**
     * Creates additional files and directories needed.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function createFiles(string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Creating additional directories and files...</info>');
        }

        $files = [
            [
                __DIR__ . '/../../bin/phpunit.template.yml',
                $directory . '/.github/workflows/phpunit.yml',
            ],
            [
                __DIR__ . '/../../bin/phpunit.template.xml',
                $directory . '/phpunit.xml.dist',
            ],
            [
                __DIR__ . '/../../output/.gitkeep',
                $directory . '/tests/_support/.gitkeep',
            ],
        ];

        try {
            foreach ($files as $file) {
                $this->fs->copy($file[0], $file[1], true);
            }
        } catch (IOExceptionInterface $e) {
            $this->output->writeln('<error>' . $e->getMessage() . '</error>');
        }

        return $this;
    }

    /**
     * Configures the system path
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function configureSystemPath(string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Configuring the system path...</info>');
        }

        $files = [
            $directory . DIRECTORY_SEPARATOR . 'app/Config/Paths.php',
            $directory . DIRECTORY_SEPARATOR . 'phpunit.xml.dist',
        ];

        foreach ($files as $file) {
            if ($this->fs->exists($file)) {
                $contents = file_get_contents($file);

                if ($this->input->getOption('dev')) {
                    $contents = str_replace(['vendor/codeigniter4/framework', '{path}'], 'vendor/codeigniter4/codeigniter4', $contents);
                    $contents = str_replace('/../../system', '/../../vendor/codeigniter4/codeigniter4/system', $contents);
                } else {
                    $contents = str_replace(['vendor/codeigniter4/codeigniter4', '{path}'], 'vendor/codeigniter4/framework', $contents);
                    $contents = str_replace('/../../system', '/../../vendor/codeigniter4/framework/system', $contents);
                }

                try {
                    $this->fs->dumpFile($file, $contents);
                } catch (IOExceptionInterface $e) {
                    $this->output->writeln('<error>' . $e->getMessage() . '</error>');
                }
            }
        }

        return $this;
    }

    /**
     * Prepares permissions on the 'writable' directory.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function prepareWritableDirectory(string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Preparing permissions on "writable" directory...</info>');
        }

        try {
            $this->fs->chmod($directory . DIRECTORY_SEPARATOR . 'writable', 0755, 0000, true);
        } catch (IOExceptionInterface $e) {
            $this->output->writeln('<error>' . $e->getMessage() . '</error>');
            $this->output->writeln('<comment>You should verify that you have permissions on the "writable" directory.</comment>');
        }

        return $this;
    }

    /**
     * Clean up the zip file
     *
     * @param string $zipFile
     *
     * @return $this
     */
    protected function cleanUp(string $zipFile)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Cleaning up the zip file...</info>');
        }

        try {
            $this->fs->chmod($zipFile, 0777);
            $this->fs->remove($zipFile);
        } catch (IOExceptionInterface $e) {
            $this->output->writeln('<error>' . $e->getMessage() . '</error>');
            $this->output->writeln('<comment>Cannot clean up the zip file. Please delete it yourself.</comment>');
        }

        return $this;
    }

    /**
     * Finds the composer executable.
     *
     * @return string
     */
    protected function findComposerPhar(): string
    {
        $composerPhar = getcwd() . DIRECTORY_SEPARATOR . 'composer.phar';
        $phpBinary    = (new PhpExecutableFinder())->find();

        if ($this->fs->exists($composerPhar)) {
            return escapeshellarg($phpBinary) . ' ' . escapeshellarg($composerPhar);
        }

        return 'composer';
    }

    /**
     * Prepares the composer.json with additional details.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function prepareComposerJson(string $directory)
    {
        if ($this->output->isVerbose()) {
            $this->output->writeln('<info>Preparing composer.json...</info>');
        }

        $composerPath = $directory . DIRECTORY_SEPARATOR . 'composer.json';
        $templateJson = json_decode(file_get_contents(__DIR__ . '/../../bin/template.json'), true);

        if (file_exists($composerPath)) {
            @unlink($composerPath);
        }

        $git = $this->getGitConfig();

        // create package name
        $name = basename(realpath($directory));
        $name = preg_replace('{(?:([a-z])([A-Z])|([A-Z])([A-Z][a-z]))}', '\\1\\3-\\2\\4', $name);
        $name = mb_strtolower($name);

        if (isset($git['github.user'])) {
            $name = $git['github.user'] . '/' . $name;
        } elseif (get_current_user()) {
            $name = get_current_user() . '/' . $name;
        } else {
            $name .= '/' . $name;
        }

        $name = mb_strtolower($name);

        // get author details
        $author = [
            'name'  => '',
            'email' => '',
        ];

        if (isset($git['user.name'])) {
            $author['name'] = $git['user.name'];
        }

        if (isset($git['user.email'])) {
            $author['email'] = $git['user.email'];
        }

        $templateJson['name']      = $name;
        $templateJson['authors'][] = $author;

        $this->fs->dumpFile($composerPath, json_encode(
            $this->appendComposerJson($templateJson),
            JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE
        ) . "\n");

        return $this;
    }

    /**
     * Appends additional details to composer.json
     *
     * @param array $composerJson
     * @param array $templateJson
     *
     * @return array
     */
    protected function appendComposerJson(array $composerJson, ?array $templateJson = null)
    {
        $templateJson  = $templateJson ?? $composerJson;

        $framework = ($this->input->getOption('dev'))
            ? ['codeigniter4/codeigniter4' => 'dev-develop']
            : ['codeigniter4/framework' => '^4'];

        // minimum stability
        $this->input->getOption('dev')
            ? $composerJson['minimum-stability'] = 'dev'
            : $composerJson['minimum-stability'] = 'stable';

        // prefer stable
        $composerJson['prefer-stable'] = isset($composerJson['prefer-stable'])
            ? true
            : $templateJson['prefer-stable'];

        // repositories
        if ($this->input->getOption('dev')) {
            if (!isset($composerJson['repositories'])) {
                $composerJson['repositories'] = [];
            }

            $composerJson['repositories'][] = [
                'type' => 'vcs',
                'url'  => 'https://github.com/codeigniter4/codeigniter4',
            ];
        }

        // require
        $composerJson['require'] = isset($composerJson['require'])
            ? array_merge($composerJson['require'], $templateJson['require'], $framework)
            : array_merge($templateJson['require'], $framework);

        // require-dev
        $composerJson['require-dev'] = isset($composerJson['require-dev'])
            ? array_merge($composerJson['require-dev'], $templateJson['require-dev'])
            : $templateJson['require-dev'];

        // autoload-dev
        $composerJson['autoload-dev']['psr-4'] = isset($composerJson['autoload-dev']['psr-4'])
            ? array_merge($composerJson['autoload-dev']['psr-4'], $templateJson['autoload-dev']['psr-4'])
            : $templateJson['autoload-dev']['psr-4'];

        // post-update-cmd
        $composerJson['scripts']['post-update-cmd'] = isset($composerJson['scripts']['post-update-cmd'])
            ? array_unique(array_merge($composerJson['scripts']['post-update-cmd'], $templateJson['scripts']['post-update-cmd']))
            : $templateJson['scripts']['post-update-cmd'];

        return $composerJson;
    }

    /**
     * Gets the location of the git binary.
     *
     * @throws RuntimeException
     * @return string
     */
    protected function getGitBinary(): string
    {
        $gitBin = (new ExecutableFinder())->find('git');

        if (null === $gitBin) {
            throw new RuntimeException('Git is not installed in your machine.');
        }

        return $gitBin;
    }

    /**
     * Gets the git config variables.
     *
     * @throws RuntimeException
     * @return array
     */
    protected function getGitConfig(): array
    {
        if ($this->gitConfig) {
            return $this->gitConfig;
        }

        $gitBin = escapeshellarg($this->getGitBinary());
        $cmd    = Process::fromShellCommandline($gitBin . ' config -l');
        $cmd->run();

        if ($cmd->isSuccessful()) {
            $this->gitConfig = [];

            preg_match_all('{^([^=]+)=(.*)$}m', $cmd->getOutput(), $matches, PREG_SET_ORDER);
            foreach ($matches as $match) {
                $this->gitConfig[$match[1]] = $match[2];
            }

            return $this->gitConfig;
        }

        return $this->gitConfig = [];
    }

    /**
     * Initializes an empty Git repository.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function initializeGit(string $directory)
    {
        if ($this->input->getOption('with-git')) {
            if ($this->output->isVerbose()) {
                $this->output->writeln("<info>Initializing empty Git repository in {$directory}...</info>");
            }

            $gitBin = escapeshellarg($this->getGitBinary());
            $cmd    = Process::fromShellCommandline($gitBin . ' init', $directory);
            $cmd->setTimeout(3600)->run();

            if ($cmd->isSuccessful()) {
                $this->output->writeln("<comment>Empty Git repository initialized at {$directory}</comment>");
            } else {
                $this->output->writeln('<error>Git initialization failed. Please run "git init" by yourself.</error>');
            }

            return $this;
        }

        return $this;
    }

    /**
     * Initializes the repository using Git Flow.
     *
     * @param string $directory
     *
     * @return $this
     */
    protected function initializeGitFlow(string $directory)
    {
        if ($this->input->getOption('with-gitflow')) {
            if ($this->output->isVerbose()) {
                $this->output->writeln('<info>Initializing Git Flow...</info>');
            }

            if (!(new ExecutableFinder())->find('git-flow')) {
                $this->output->writeln('<error>Git Flow is not installed in your machine.</error>');
                return $this;
            }

            $gitBin   = escapeshellarg($this->getGitBinary());
            $commands = [
                $gitBin . ' init',
                $gitBin . ' flow init -d -f --local -t v',
            ];

            $cmd = Process::fromShellCommandline(implode(' && ', $commands), $directory);
            $cmd->setTimeout(3600)->run();

            if ($cmd->isSuccessful()) {
                $this->output->writeln("<comment>Git Flow initialized at {$directory}</comment>");
            } else {
                $this->output->writeln('<error>Git Flow initialization failed. Please run "git flow init" by yourself.</error>');
            }

            return $this;
        }

        return $this;
    }

    /**
     * The main installation logic.
     *
     * @param string $directory
     *
     * @return int
     */
    protected function installApplication(string $directory)
    {
        $composer = $this->findComposerPhar();
        $commands = [
            $composer . ' install --ansi',
        ];

        if ($this->input->getOption('no-ansi')) {
            $commands = array_map(function ($command) {
                return str_replace('--ansi', '--no-ansi', $command);
            }, $commands);
        }

        if ($this->input->getOption('quiet')) {
            $commands = array_map(function ($command) {
                return $command . ' --quiet';
            }, $commands);
        }

        $cmd = Process::fromShellCommandline(implode(' && ', $commands), $directory);

        if ('\\' !== DIRECTORY_SEPARATOR && file_exists('/dev/tty') && is_readable('/dev/tty')) {
            try {
                $cmd->setTty(true);
            } catch (RuntimeException $e) {
                $this->output->writeln('<comment>Warning: ' . $e->getMessage() . '</comment>');
            }
        }

        $output = $this->output;

        $cmd
            ->setTimeout(3600)
            ->run(function ($type, $line) use ($output) {
                $output->write($line);
            })
        ;

        if ($cmd->isSuccessful()) {
            $this->output->writeln('<comment>Application ready! Start building your craft now!</comment>');
            return 0;
        }

        $this->output->writeln('<error>Application scaffolding failed.</error>');
        return 1;
    }
}
